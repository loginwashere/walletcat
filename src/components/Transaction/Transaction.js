import React, { Component, PropTypes } from 'react';
import { Badge } from 'react-bootstrap';
import format from 'date-fns/format';

import './Transaction.css';

export default class Transaction extends Component {
  render() {
    const { transaction, fromAccount, toAccount, category } = this.props;
    const date = format(new Date(transaction.date), 'YYYY-MM-DD HH:mm:ss')
    return (
      <li className="list-group-item">
        <p>
          <Badge bsStyle="danger" pullRight>{transaction.fromAmount}</Badge>
          From {fromAccount.name}
        </p>
        <p>
          <Badge bsStyle="success" pullRight>{transaction.toAmount}</Badge>
          To {toAccount.name}
        </p>
        <p className="list-group-item-text">
          <Badge pullRight>{date}</Badge>
          {category && category.name}
        </p>
        {transaction.description && <p>{transaction.description}</p>}
      </li>
    );
  }
}

Transaction.propTypes = {
  transaction: PropTypes.shape({
    id: PropTypes.number.isRequired,
    description: PropTypes.string,
    date: PropTypes.string.isRequired,
    fromAmount: PropTypes.number.isRequired,
    toAmount: PropTypes.number.isRequired
  }).isRequired,
  fromAccount: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired
  }).isRequired,
  toAccount: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired
  }).isRequired,
  category: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired
  }).isRequired
}
