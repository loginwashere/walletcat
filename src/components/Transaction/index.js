import React, { Component, PropTypes } from 'react'
import { Link } from 'react-router'
import { Badge, Row, Col } from 'react-bootstrap'
import format from 'date-fns/format'

import './style.less'

const TransactionItem = ({ transactionItem, account, currency, type, first }) => (
  <Row>
    {first && <Col sm={1} xs={3}>
      {type === 'credit'
        ? 'From'
        : 'To'}
    </Col>}
    <Col sm={9}
         smOffset={!first
          ? 1
          : 0}
         xs={7}
         xsOffset={!first
          ? 3
          : 9}><b>{account.name}</b></Col>
    <Col sm={2} xs={2}>
      <Badge bsStyle={type === 'credit'
              ? 'danger'
              : 'success'}
             pullRight>
        {transactionItem.amount} {currency.name}
      </Badge>
    </Col>
  </Row>
)

TransactionItem.propTypes = {
  transactionItem: PropTypes.shape({
    amount: PropTypes.string.isRequired
  }).isRequired,
  account: PropTypes.shape({
    name: PropTypes.string.isRequired
  }).isRequired,
  currency: PropTypes.shape({
    name: PropTypes.string.isRequired
  }).isRequired,
  type: PropTypes.string.isRequired,
  first: PropTypes.bool
}

const TransactionItemList = ({
  transactionItemIds,
  transactionItems,
  accounts,
  userCurrencies,
  currencies,
  className,
  type
}) => (
  <div className={className}>
    {transactionItemIds.map((transactionItemId, index) => {
      const transactionItem = transactionItems[transactionItemId]
      const account = transactionItem && accounts[transactionItem.accountId]
      const userCurrency = account && userCurrencies[account.currencyId]
      const currency = userCurrency && currencies[userCurrency.currencyId]
      return (
        transactionItem &&
        account &&
        currency &&
        <TransactionItem key={transactionItem.id}
                         transactionItem={transactionItem}
                         account={account}
                         currency={currency}
                         type={type}
                         first={index === 0} />
      )
    })}
  </div>
)

TransactionItemList.propTypes = {
  transactionItemIds: PropTypes.array.isRequired,
  transactionItems: PropTypes.object,
  accounts: PropTypes.object,
  userCurrencies: PropTypes.object,
  currencies: PropTypes.object,
  className: PropTypes.string,
  type: PropTypes.string.isRequired
}

export class Transaction extends Component {
  render() {
    const {
      transaction,
      accounts,
      transactionItems,
      userCurrencies,
      currencies,
      category
    } = this.props
    const debitTransactionItemIds = transaction.transactionItems
      .filter(transactionItemId => transactionItems[transactionItemId] &&
        transactionItems[transactionItemId].type === 'debit')
    const creditTransactionItemIds = transaction.transactionItems
      .filter(transactionItemId => transactionItems[transactionItemId] &&
        transactionItems[transactionItemId].type === 'credit')
    const date = format(new Date(transaction.date), 'YYYY-MM-DD HH:mm:ss')
    return (
      <Link to={`/transactions/${transaction.id}`}
            className="list-group-item">
        <TransactionItemList className="list-group-item-text truncate"
                             transactionItemIds={creditTransactionItemIds}
                             accounts={accounts}
                             userCurrencies={userCurrencies}
                             currencies={currencies}
                             transactionItems={transactionItems}
                             type={'credit'} />
        <TransactionItemList className="list-group-item-text truncate"
                             transactionItemIds={debitTransactionItemIds}
                             accounts={accounts}
                             userCurrencies={userCurrencies}
                             currencies={currencies}
                             transactionItems={transactionItems}
                             type={'debit'} />
        <div className="list-group-item-text truncate">
          <Row>
            <Col sm={1} xs={3}>Category</Col>
            <Col sm={9} xs={6}><b>{category.name}</b></Col>
            <Col sm={2} xs={3}>
              <Badge pullRight>{date}</Badge>
            </Col>
          </Row>
        </div>
        {transaction.description && <div className="list-group-item-text truncate">
            <Row>
              <Col sm={1} xs={12}>Description</Col>
              <Col sm={11} xs={12}>{transaction.description}</Col>
            </Row>
          </div>}
      </Link>
    )
  }
}

Transaction.propTypes = {
  transaction: PropTypes.shape({
    id: PropTypes.string.isRequired,
    description: PropTypes.string,
    date: PropTypes.string.isRequired,
    fromAmount: PropTypes.string,
    toAmount: PropTypes.string
  }).isRequired,
  accounts: PropTypes.object,
  transactionItems: PropTypes.object,
  userCurrencies: PropTypes.object,
  currencies: PropTypes.object,
  category: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired
  }).isRequired
}

export default Transaction
