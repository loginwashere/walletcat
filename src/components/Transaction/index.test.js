import React from 'react'
import renderer from 'react-test-renderer'
import Transaction from '.'
import {
  transactionSeeder,
  transactionItemSeeder,
  categorySeeder,
  accountSeeder,
  userCurrencySeeder,
  currencySeeder
} from '../../../server/seeds'

describe('components:Transaction:', () => {
  it('renders without crashing', () => {
    const transaction = transactionSeeder.items[0]
    const transactionItems = transactionItemSeeder.items
      .reduce((obj, item) => ({ ...obj, [item.id]: item }), {})
    const testTransaction = {
      ...transaction,
      transactionItems: Object.keys(transactionItems)
        .filter(transactionItemId => transactionItems[transactionItemId].transactionId === transaction.id)
    }
    const category = categorySeeder.items.filter(item => item.id === transaction.categoryId)[0]
    const accounts = accountSeeder.items
      .reduce((obj, item) => ({ ...obj, [item.id]: item }), {})
    const userCurrencies = userCurrencySeeder.items
      .reduce((obj, item) => ({ ...obj, [item.id]: item }), {})
    const currencies = currencySeeder.items
      .reduce((obj, item) => ({ ...obj, [item.id]: item }), {})

    const tree = renderer.create(
      <Transaction transaction={testTransaction}
                   accounts={accounts}
                   transactionItems={transactionItems}
                   userCurrencies={userCurrencies}
                   currencies={currencies}
                   category={category} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
