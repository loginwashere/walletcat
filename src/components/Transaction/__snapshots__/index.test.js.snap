exports[`components:Transaction: renders without crashing 1`] = `
<a
  className="list-group-item"
  onClick={[Function]}
  style={Object {}}>
  <div
    className="list-group-item-text truncate">
    <div
      className="row">
      <div
        className="col-sm-1 col-xs-3">
        From
      </div>
      <div
        className="col-sm-9 col-sm-offset-0 col-xs-7 col-xs-offset-9">
        <b>
          Wallet
        </b>
      </div>
      <div
        className="col-sm-2 col-xs-2">
        <span
          className="badge badge-danger pull-right">
          100.00000000
           
          AED
        </span>
      </div>
    </div>
  </div>
  <div
    className="list-group-item-text truncate">
    <div
      className="row">
      <div
        className="col-sm-1 col-xs-3">
        To
      </div>
      <div
        className="col-sm-9 col-sm-offset-0 col-xs-7 col-xs-offset-9">
        <b>
          Stash
        </b>
      </div>
      <div
        className="col-sm-2 col-xs-2">
        <span
          className="badge badge-success pull-right">
          100.00000000
           
          AED
        </span>
      </div>
    </div>
  </div>
  <div
    className="list-group-item-text truncate">
    <div
      className="row">
      <div
        className="col-sm-1 col-xs-3">
        Category
      </div>
      <div
        className="col-sm-9 col-xs-6">
        <b>
          Fast food
        </b>
      </div>
      <div
        className="col-sm-2 col-xs-3">
        <span
          className="badge pull-right">
          1995-02-10 20:45:43
        </span>
      </div>
    </div>
  </div>
  <div
    className="list-group-item-text truncate">
    <div
      className="row">
      <div
        className="col-sm-1 col-xs-12">
        Description
      </div>
      <div
        className="col-sm-11 col-xs-12">
        description
      </div>
    </div>
  </div>
</a>
`;
